govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/coherence/metric.py

```diff
@@ -0,0 +1,193 @@
+    "STEP 9 - Final Coherence-Only Check: Confirm your evaluation focuses solely on: structure, clarity, consistency, perspective, and logical presentation. You are NOT evaluating: correctness, completeness, relevance, or politeness. An answer can say 'I don't know' and score a 5 if it's clear about what it does/doesn't know. An answer can be comprehensive and score a 1 if it contradicts itself.",
+    "STEP 10 - Score Assignment with Justification: Based on Steps 1-9, assign a score using this logic: (a) ANY contradiction on core actions/obligations = 1 or 2, (b) Wrong/mixed perspective or buried critical context = 2 or 3, (c) Notable ambiguity or flow issues = 3 or 4, (d) Minor inefficiency only = 4, (e) No issues detected = 5. Provide specific evidence from the answer for your score. Default to the LOWER score if uncertain between two.",
+]
+
+
+class CoherenceMetric(BaseMetric):
+    _required_params: list[LLMTestCaseParams] = [
+        LLMTestCaseParams.INPUT,
+        LLMTestCaseParams.ACTUAL_OUTPUT,
+    ]
+    evaluation_template: Type[CoherenceTemplate] = CoherenceTemplate
+
+    def __init__(
+        self,
+        model: DeepEvalBaseLLM | str | None,
+        threshold: float = DEFAULT_COHERENCE_THRESHOLD,
+        include_reason: bool = True,
+        strict_mode: bool = False,
+        verbose_mode: bool = False,
+        evaluation_template: Type[CoherenceTemplate] = CoherenceTemplate,
+    ):
+        self.threshold = 1.0 if strict_mode else threshold
+        self.include_reason = include_reason
+        self.strict_mode = strict_mode
+        self.verbose_mode = verbose_mode
+        self.evaluation_template = evaluation_template
+
+        self.evaluation_steps = COHERENCE_EVALUATION_STEPS
+        self.rubric = COHERENCE_RUBRIC
+        self.score_range = COHERENCE_SCORE_RANGE
+        self.additional_context = COHERENCE_ADDITIONAL_CONTEXT
+
+        self.model, self.using_native_model = initialize_model(model)
+        self.evaluation_model = self.model.get_model_name()
+        self.evaluation_cost: float | None = None
+
+        self.score: float | None = None
+        self.reason: str | None = None
+        self.raw_score: int | None = None
+
+    async def a_measure(
+        self,
+        test_case: LLMTestCase,
+        _show_indicator: bool = True,
+        _in_component: bool = False,
+    ) -> float:
+        check_llm_test_case_params(test_case, self._required_params, self)
+
+        if self.using_native_model:
+            self.evaluation_cost = 0.0
+        else:
+            self.evaluation_cost = None
+
+        with metric_progress_indicator(
+            self,
+            async_mode=True,
+            _show_indicator=_show_indicator,
+            _in_component=_in_component,
+        ):
+            self.error = None
+            prompt = self.evaluation_template.evaluate(
+                user_input=test_case.input or "",
+                actual_output=test_case.actual_output or "",
```

Comment:
I'd have thought we don't need the fallbacks here as we have these fields flagged as required input?